src/bootstrap/native.rs

```diff
@@ -111,6 +116,275 @@ pub fn prebuilt_llvm_config(
+    let nix_deps_dir = NIX_DEPS_DIR.get_or_init(|| {
+        // Run `nix-build` to "build" each dependency (which will likely reuse
+        // the existing `/nix/store` copy, or at most download a pre-built copy).
+        //
+        // Importantly, we create a gc-root called `.nix-deps` in the `build/`
+        // directory, but still reference the actual `/nix/store` path in the rpath
+        // as it makes it significantly more robust against changes to the location of
+        // the `.nix-deps` location.
+        //
+        // bintools: Needed for the path of `ld-linux.so` (via `nix-support/dynamic-linker`).
+        // zlib: Needed as a system dependency of `libLLVM-*.so`.
+        // patchelf: Needed for patching ELF binaries (see doc comment above).
+        let nix_deps_dir = t!(builder.out.join(".nix-deps").canonicalize());
+        const NIX_EXPR: &str = "
+        with (import <nixpkgs> {});
+        symlinkJoin {
+            name = \"rust-stage0-dependencies\";
+            paths = [
+            zlib
+            patchelf
+            stdenv.cc.bintools
+            ];
+        }
+        ";
+        // FIXME: support non-utf8 paths?
+        let nix_deps_dir_str = nix_deps_dir.to_str().unwrap();
+        nix_build_succeeded = builder.try_run(Command::new("nix-build").args(&[
+            "-E",
+            NIX_EXPR,
+            "-o",
+            nix_deps_dir_str,
+        ]));
+        nix_deps_dir_str.to_string()
+    });
+    if !nix_build_succeeded {
+        return;
+    }
+
+    let mut patchelf = Command::new(Path::new(nix_deps_dir).join("bin/patchelf"));
+    let rpath_entries = format!(
+        // ORIGIN is a relative default, all binary and dynamic libraries we ship
+        // appear to have this (even when `../lib` is redundant).
+        "$ORIGIN/../lib:{}/lib",
+        nix_deps_dir
+    );
+    patchelf.args(&["--set-rpath", &rpath_entries]);
+    if !fname.ends_with(".so") {
+        // Finally, set the corret .interp for binaries
+        let dynamic_linker_path = Path::new(nix_deps_dir).join("nix-support/dynamic-linker");
+        let dynamic_linker = t!(String::from_utf8(t!(fs::read(dynamic_linker_path))));
+        patchelf.args(&["--set-interpreter", dynamic_linker.trim_end()]);
+    }
+
+    builder.try_run(&mut patchelf);
+}
+
+fn download_component(builder: &Builder<'_>, base: &str, url: &str, dest_path: &Path) {
+    // Use a temporary file in case we crash while downloading, to avoid a corrupt download in cache/.
+    let tempfile = t!(tempfile::NamedTempFile::new());
+    let temppath = tempfile.path().to_owned();
+    drop(tempfile);
+    let tempfile_str = temppath.to_str().expect("tempdir must be valid unicode");
+    // FIXME: support `do_verify` (only really needed for nightly rustfmt)
+    download_with_retries(builder, tempfile_str, &format!("{}/{}", base, url));
+    builder.verbose(&format!("moving {} to {}", tempfile_str, dest_path.display()));
+    t!(fs::rename(temppath, dest_path));
```

Comment:
It turns out that python is smarter than rust about this:
> If the destination is on the current filesystem, then [os.rename()](https://docs.python.org/3/library/os.html#os.rename) is used. Otherwise, src is copied to dst using copy_function and then removed. In case of symlinks, a new symlink pointing to the target of src will be created in or as dst and src will be removed.

https://docs.python.org/3/library/shutil.html
@m-ou-se do you know if it's possible to do the same in libstd, instead of giving a hard error when the source and destination are on different file systems?

In the meantime I'll replicate python's logic in bootstrap.

Reply:
We have a copy function on `Build` that tries to hardlink and otherwise copies the raw bytes. You can put the rename function you write next to it.